clippy_lints/src/loops/manual_find.rs

```diff
@@ -0,0 +1,189 @@
+use super::utils::make_iterator_snippet;
+use super::MANUAL_FIND;
+use crate::rustc_hir::intravisit::Visitor;
+use clippy_utils::{
+    diagnostics::span_lint_and_sugg,
+    higher, peel_blocks_with_stmt,
+    source::snippet_with_applicability,
+    ty::is_type_diagnostic_item,
+    visitors::{expr_visitor, find_all_ret_expressions},
+};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{BindingAnnotation, Expr, ExprKind, ItemKind, Node, Pat, PatKind, QPath, StmtKind};
+use rustc_lint::LateContext;
+use rustc_span::source_map::Span;
+use rustc_span::sym;
+use rustc_span::symbol::Ident;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    pat: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    body: &'tcx Expr<'_>,
+    span: Span,
+) {
+    let inner_expr = peel_blocks_with_stmt(body);
+    if let Some(higher::If {
+        cond,
+        then,
+        r#else: None,
+    }) = higher::If::hir(inner_expr)
+    {
+        let mut pat_expr = None;
+        let mut ident_collector = PatIdentCollector::default();
+        ident_collector.visit_pat(pat);
+        if let [ident] = &ident_collector.idents[..] {
+            if ident_in_expr(cx, *ident, cond) {
+                pat_expr = Some(*ident);
+            }
+        }
+        // Check for the specific case that the result is returned and optimize suggestion for that (more
+        // cases can be added later)
+        if_chain! {
+            if let Some(ident) = pat_expr;
+            if let ExprKind::Block(block, _) = then.kind;
+            if let [stmt] = block.stmts; // For now, only deal with the case w/out side effects/mapping
+            if let StmtKind::Semi(semi) = stmt.kind;
+            if let ExprKind::Ret(Some(ret_value)) = semi.kind;
+            let ret_ty = cx.typeck_results().expr_ty(ret_value);
+            if is_type_diagnostic_item(cx, ret_ty, sym::Option);
+            if let ExprKind::Call(_, [inner_ret]) = ret_value.kind;
+            if let Some(i) = ident_of(inner_ret);
+            if i == ident;
+            if let Node::Item(parent_fn)
+                = cx.tcx.hir().get_by_def_id(cx.tcx.hir().get_parent_item(body.hir_id));
+            if let ItemKind::Fn(_, _, body_id) = parent_fn.kind;
+            let parent_fn = &cx.tcx.hir().body(body_id).value;
+            let ret_exprs = collect_ret_exprs(cx, parent_fn);
+            if ret_exprs.len() == 2;
+            // Some(binding)
+            if let ExprKind::Call(_, [binding]) = &ret_exprs[0].kind;
+            // None
+            if let ExprKind::Path(..) = &ret_exprs[1].kind;
```

Comment:
Checking the ret values runs into trouble with

```rust
fn for_in_condition(a: &[i32], b: bool) -> Option<i32> {
    if b {
        for &x in a {
            if x == 1 {
                return Some(x);
            }
        }
    }
    None
}
```
```rust
fn intermediate_statements(a: &[i32]) -> Option<i32> {
    for &x in a {
        if x == 1 {
            return Some(x);
        }
    }

    println!("side effect");

    None
}
```

I can't think of a pleasant way to check this other than mechanically using [`parent_iter`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.parent_iter). Passing `expr` into `manual_kind::check` from `mod.rs`, and calling `parent_iter` on that should give you:

- A `Stmt`
- A `Block`   
  - The above statement should be the last one of the block
  - The [`block.expr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.Block.html#structfield.expr) should be `Some(Expr { ExprKind::Path(path), .. })` where `is_lang_ctor(cx, path, OptionNone)`
- A fn-like thing, you can check [`Node::fn_sig`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.Node.html#method.fn_sig) for this

there may be some extra `Expr`s mixed in like `ExprKind::Block` you have to skip over. You'd probably want to stash all this in a different function

Reply:
This works great, thanks! I've run into some trouble with the `fn_sig` in the case that the find implementation occurs within a closure, but it seems sufficient with the tests that I have to just check the `Stmt` and `Block` like you described. Is there an edge case that I'm not considering that means we have to check for a `fn_sig` as well?

Reply:
Ah whoops, I didn't know closures have no `FnSig`. Looks like [`fn_kind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.Node.html#method.fn_kind) instead would do the trick

Here's one I think would be a FP without such a check:

```rust
fn in_block(a: &[i32]) -> Option<i32> {
    let should_be_none = {
        for &x in a {
            if x == 1 {
                return Some(x);
            }
        }
        None
    };
    
    assert!(should_be_none.is_none());
    
    should_be_none
}
```

Transforming it into a find would remove the early return